Changes for RDMA with VMM

dayatsin-amd opened the pull request:
Adding code to retrieve the mmap_offset and then create a CPU mapping with PROT_NONE. This allows the KFD peerdirect drivers to look up the entry for the mapping by virtual address.

atgutier reviewed (commented):
Overall API changes & additions look good to me. Would only have minor stylistic recommendations now so will wait until this is out of WIP before adding any additional comments.

dayatsin-amd commented:
@atgutier  Thanks Tony. I am done. This is ready for review

atgutier commented on projects/rocr-runtime/libhsakmt/include/hsakmt/hsakmt.h:
```diff
@@ -439,6 +439,17 @@ hsaKmtAllocMemoryAlign(
+    HSAuint64       SizeInBytes,            //IN  (multiple of page size)
+    HSAuint64       Alignment,              //IN  (power of 2 and >= page size)
+    HsaMemFlags     MemFlags,               //IN
+    void**          MemoryAddress,           //IN/OUT (page-aligned)
```
Minor style nit: can you align the comments?

dayatsin-amd replied:
ACK

atgutier commented on projects/rocr-runtime/libhsakmt/src/fmm.h:
```diff
@@ -51,7 +51,7 @@ void hsakmt_fmm_destroy_process_apertures(void);
 /* Memory interface */
 void *hsakmt_fmm_allocate_scratch(uint32_t gpu_id, void *address, uint64_t MemorySizeInBytes);
-void *hsakmt_fmm_allocate_device(uint32_t gpu_id, uint32_t node_id, void *address,
+void *hsakmt_fmm_allocate_device(uint32_t gpu_id, uint32_t node_id, void *address, uint64_t *offset,
```
Should this be mmap_offset to be consistent with the impl and more precise?

dayatsin-amd replied:
ACK

atgutier commented on projects/rocr-runtime/runtime/hsa-runtime/core/inc/driver.h:
```diff
@@ -142,7 +142,7 @@ class Driver {
   virtual hsa_status_t AllocateMemory(const MemoryRegion &mem_region,
                                       MemoryRegion::AllocateFlags alloc_flags,
-                                      void **mem, size_t size,
+                                      void **mem, uint64_t *offset, size_t size,
```
Can you update the doxygen comments to include this parameter?

dayatsin-amd replied:
ACK

atgutier commented on projects/rocr-runtime/runtime/hsa-runtime/core/inc/runtime.h:
```diff
@@ -885,6 +886,7 @@ class Runtime {
     size_t size;
     int drm_fd;
     void* drm_cpu_addr;  // CPU Buffer address
+    void* cpu_mapping; //DYSDEBUG
```
This is marked as debug, should it be here in the final change? I don't see it used anywhere either other than being mapped on creation and unmapped on destruction.

dayatsin-amd replied:
It is final change. I forgot to remove my comment :)

atgutier commented on projects/rocr-runtime/libhsakmt/include/hsakmt/hsakmt.h:
```diff
@@ -439,6 +439,17 @@ hsaKmtAllocMemoryAlign(
+    HSAuint64       Alignment,              //IN  (power of 2 and >= page size)
+    HsaMemFlags     MemFlags,               //IN
+    void**          MemoryAddress,           //IN/OUT (page-aligned)
+    HSAuint64*      Offset
```
Should this be called MmapOffset or similar throughout to be more consistent/precise?

kentrussell commented on projects/rocr-runtime/libhsakmt/src/memory.c:
```diff
@@ -213,7 +225,12 @@ HSAKMT_STATUS HSAKMTAPI hsaKmtAllocMemoryAlign(HSAuint32 PreferredNode,
+		*MmapOffset = offset;
+
+
+	pr_debug("[%s] node %d address %p size %lu offset:%lu from device\n",
```
Maybe "offset %lu" to match the rest of the comment. In case people are grepping/parsing. Such a nitpick, but since you need to rebase anyways, I don't feel so bad saying it

dayatsin-amd replied:
We need the VA to be able to call mmap inside the MappedHandle constructor here:

cpu_mapping = mmap(**va**, size, PROT_NONE, MAP_SHARED | MAP_FIXED,
                drm_fd, reinterpret_cast<uint64_t>(mem_handle->mmap_offset));
                
Before, it was only the key in mapped_handle_map_ that was used to track the VA. But when we are in the MappedHandle constructor, we do not have the key.

dayatsin-amd replied:
ACK

kentrussell commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:
```diff
@@ -3392,7 +3394,7 @@ hsa_status_t Runtime::VMemoryHandleMap(void* va, size_t size, size_t in_offset,
   mapped_handle_map_.emplace(
       std::piecewise_construct, std::forward_as_tuple(va),
-      std::forward_as_tuple(&memoryHandleIt->second, addressHandle, offset, size, drm_fd,
+      std::forward_as_tuple(&memoryHandleIt->second, addressHandle, va, export_offset, size, drm_fd,
```
Pleading ignorance here. Why is va added now?

dayatsin-amd replied:
We need the VA to be able to call mmap inside the MappedHandle constructor here:

cpu_mapping = mmap(**va**, size, PROT_NONE, MAP_SHARED | MAP_FIXED,
                drm_fd, reinterpret_cast<uint64_t>(mem_handle->mmap_offset));
                
Before, it was only the key in mapped_handle_map_ that was used to track the VA. But when we are in the MappedHandle constructor, we do not have the key.

dayatsin-amd committed:
rocr: Add support for VMM and RDMA

Add extra CPU mapping so that kernel-mode drivers can look up the memory
mapping by virtual address.

ypapadop-amd commented on projects/rocr-runtime/runtime/hsa-runtime/core/inc/runtime.h:
```diff
@@ -879,12 +879,11 @@ class Runtime {
 
   struct MappedHandle {
-    MappedHandle(MemoryHandle *mem_handle, AddressHandle *address_handle,
+    public:
```
```suggestion
```
Default visibility in struct is public.

dayatsin-amd replied:
ACK

dayatsin-amd replied:
ACK

ypapadop-amd commented on projects/rocr-runtime/runtime/hsa-runtime/core/inc/runtime.h:
```diff
@@ -879,12 +879,11 @@ class Runtime {
   struct MappedHandle {
-    MappedHandle(MemoryHandle *mem_handle, AddressHandle *address_handle,
+    public:
+    MappedHandle(MemoryHandle *mem_handle, AddressHandle *address_handle, void* va,
```
Pointer alignment seems off here.

dayatsin-amd replied:
ACK

ypapadop-amd commented on projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp:
```diff
@@ -3595,17 +3595,46 @@ hsa_status_t Runtime::MappedHandleAllowedAgent::EnableAccess(hsa_access_permissi
+    shareable_handle(shareable_handle)
+{
+  /* Create a CPU mapping with PROT_NONE */
+  auto cpu_agent = ((AMD::GpuAgent*)agentOwner())->GetNearestCpuAgent();
```
```suggestion
  auto cpu_agent = static_cast<AMD::GpuAgent*>(agentOwner())->GetNearestCpuAgent();
```

dayatsin-amd replied:
ACK

ypapadop-amd commented on projects/rocr-runtime/runtime/hsa-runtime/core/inc/runtime.h:
```diff
@@ -879,12 +879,11 @@ class Runtime {
-          offset(offset), size(size), drm_fd(drm_fd),
-          drm_cpu_addr(drm_cpu_addr), shareable_handle(shareable_handle) {}
+                 hsa_access_permission_t perm, ShareableHandle shareable_handle);
+     ~MappedHandle();
```
```suggestion
```
appears to be empty below.

cfreeamd committed:
rocr: Honor uncache flag in memory_lock_to_pool()

Also, combined several flag options used in apis into a
single integer.

Signed-off-by: Chris Freehill <cfreehil@amd.com>

dayatsin-amd committed:
Update projects/rocr-runtime/runtime/hsa-runtime/core/runtime/runtime.cpp

Co-authored-by: Yiannis Papadopoulos <102817138+ypapadop-amd@users.noreply.github.com>

dayatsin-amd committed:
Update projects/rocr-runtime/runtime/hsa-runtime/core/inc/runtime.h

Co-authored-by: Yiannis Papadopoulos <102817138+ypapadop-amd@users.noreply.github.com>

dayatsin-amd commented:
/AzurePipelines run rocm-ci-caller

dayatsin-amd commented:
/AzurePipelines run rocm-ci-caller

dayatsin-amd committed:
rocr: Fix hsa_amd_pointer_info on CPU agents

Fix hsa_amd_pointer_info query returning allowd on VMM pointers for CPU
agents when CPU mapping was mapped with PROT_NONE.

cfreeamd committed:
Merge branch 'develop' into users/dayatsin/vmm-rdma
